docs(internals): data storage (redux) - expand stub (#21689)
gatsbyjs/gatsby: docs/docs/data-storage-redux.md

---
title: Data Storage (Redux)
---

During Gatsby's bootstrap & build phases, the state is stored and manipulated using the [Redux](https://redux.js.org/) library. The key purpose of using Redux in Gatsby internals is to centralize all of the state logic. Reviewing the Gatsby [reducers](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby/src/redux/reducers) and [actions](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby/src/redux/actions) folders gives a comprehensive picture of what state manipulations are possible.

## Store

The namespaces in Gatsby's Redux store are a great overview of the Gatsby internals, here are a few:

- **[Nodes](/docs/node-interface/)** - All data that’s added to Gatsby is modeled using nodes. Nodes are most commonly added by Source plugins such as `gatsby-source-filesystem`.
- **[Schema](/docs/schema-generation/)** - GraphQL schema inferred from Nodes, available for querying by page and static queries.
- **[Pages](/docs/gatsby-internals-terminology/#page)** - A `Map` of page paths to page objects. Page objects contain information needed to render a page such as component file path, page query and context.

> - **[Components](/docs/gatsby-internals-terminology/#component)** - A `Map` of component file paths to page objects.

Author: Hah yeh it reads exactly like a typo, it is correct though, it's a paraphrase of what's in the [terminology section](https://www.gatsbyjs.org/docs/gatsby-internals-terminology/#redux-components-namespace). It is confusing just sittin' there with no elaboration.

Ohh boy, had a source code dive and found that the component's version of the page object really has the juice. The page object is [pretty static](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/redux/reducers/pages.js) but the component object can reference many pages and is kept up to date by a [state machine during query extraction](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/redux/machines/page-component.js).

What do you think about a little extra info about that:

```suggestion
- **[Components](/docs/gatsby-internals-terminology/#component)** - A `Map` of component file paths to component objects. A component object is similar to a page object but can reference many page paths and is updated through the [query extraction](/docs/query-extraction/) process.
```

- **[Static Query Components](/docs/static-vs-normal-queries/#keeping-track-of-site-queries-during-build-in-redux-stores)** - A `Map` of any components detected with a static query.
- **Jobs** - Long-running processes, generally started as a side-effect to a GraphQL query. Gatsby doesn’t finish its process until all jobs are ended.

> - **Webpack** - Config for the [Webpack](https://webpack.js.org/) bundler which handles code optimization and splitting of delivered javascript bundles.

Reviewer: ```suggestion
- **Webpack** - Config for the [Webpack](/docs/webpack-and-ssr/) bundler which handles code optimization and splitting of delivered javascript bundles.
```
It's nice to link to docs on the Gatsbyjs.org site wherever possible, do you think this conceptual guide is appropriate? https://gatsbyjs.org/docs/webpack-and-ssr/

Author: Awww yiss! Much better

> The Gatsby [/redux index file](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby/src/redux/index.ts) has two key exports, `store` and `emitter`. Throughout the bootstrap and build phases, `store` is used to get the current state and dispatch actions, while `emitter` is used to register listeners for particular actions. The store is also made available to Gatsby users through the [Node APIs](/docs/node-apis/).

Reviewer: ```suggestion
The Gatsby [redux index file](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby/src/redux/index.ts) has two key exports, `store` and `emitter`. Throughout the bootstrap and build phases, `store` is used to get the current state and dispatch actions, while `emitter` is used to register listeners for particular actions. The store is also made available to Gatsby users through the [Node APIs](/docs/node-apis/).
```
The slash in `/redux` reads a bit strange in a sentence, I'd recommend omitting it or marking it as code.

## Actions

Actions dispatched in the store cause state changes through the reducers and also trigger listeners registered for that action on a [mitt](https://www.npmjs.com/package/mitt) `emitter`. While the `subscribe` Redux store method is typically used to connect a web framework like React, Gatsby only uses the `subscribe` method to connect the `emitter`.

The [Gatsby actions](/docs/actions/) are all either internal, public or restricted. The public actions, and a context relevant subset of the restricted actions, are available to users through the [Node APIs](/docs/node-apis/).

> ### Action Journey

Reviewer: ```suggestion
### Example action journey for `createRedirect`
```

> Gatsby actions have a similar journey through defining, exposing and dispatching. Let's follow the [createRedirect](/docs/actions/#createRedirect) public action:

Reviewer: ```suggestion
Gatsby actions have a similar journey through defining, exposing and dispatching. This section follows the [createRedirect](/docs/actions/#createRedirect) public action:
```
Updating to follow the Gatsby Style Guide for pronouns.

- **Side effect** - An `emitter` listener is registered for the `CREATE_REDIRECT` action type.

- **Action creator** - An action creator, `createRedirect`, is defined in the public actions file. The action has a payload, the information needed to complete the action, and a type, the string that identifies this particular action.

- **Expose bound action creator** - `createRedirect` is one of the public actions made available to all of the [Node APIs](/docs/node-apis/). A collection of public actions and the restricted actions available to the called API are bound to the Redux store dispatch. The bound action collection is then passed when calling the user's API function.

- **Dispatch** - Here is an example of the `createRedirect` call that a Gatsby user could make with the [createPages](/docs/node-apis/#createPages) API in their project's [gatsby-node.js](/docs/api-files-gatsby-node/) file:

```javascript:title=gatsby-node.js
module.exports = {
  createPages: ({ actions }) => {
    const { createRedirect } = actions
    createRedirect({
      fromPath: "/legacy-path",
      toPath: "/current-path",
    })
  },
}

> ```

Reviewer: ````suggestion
```

By walking through an action scenario in detail, you can hopefully understand more about Gatsby's internals using Redux.

## Additional resources

- [Behind the Scenes: What Makes Gatsby Great](/blog/2019-04-02-behind-the-scenes-what-makes-gatsby-great/)
- [Gatsby Magic](/docs/gatsby-magic/)
````

Upon reading the doc, it seemed like it needed a conclusion and some further reading. Let me know what you think!

Author: Agree on the conclusion sentence 👍

Hesitated on the additional resources though, reckon the best additional resource is the wider internals section that this doc is part of. They likely aready know that though, and these 2 pages are great and provide alternate angles on the internals.

---
title: Data Storage (Redux)
---

> During Gatsby's bootstrap & build phases, the state is stored and manipulated using the [Redux](https://redux.js.org/) library. The key purpose of using Redux is to centralise all of the state logic. Reviewing the Gatsby [reducers](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/reducers) and [actions](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/actions) folders gives a comprehensive picture of what state manipulations are possible.

Reviewer 2: ```suggestion
During Gatsby's bootstrap & build phases, the state is stored and manipulated using the [Redux](https://redux.js.org/) library. The key purpose of using Redux is to centralize all of the state logic. Reviewing the Gatsby [reducers](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/reducers) and [actions](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/actions) folders gives a comprehensive picture of what state manipulations are possible.
```

Author: I can't tell what the change is in this line, am I missing something in the Github UI? 🕵️‍♂️

Reviewer 2: "centralise" is a typo, so the edit is to make it "centralize". Hard to see!

> During Gatsby's bootstrap & build phases, the state is stored and manipulated using the [Redux](https://redux.js.org/) library. The key purpose of using Redux is to centralize all of the state logic. Reviewing the Gatsby [reducers](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/reducers) and [actions](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/actions) folders gives a comprehensive picture of what state manipulations are possible.

Reviewer: ```suggestion
During Gatsby's bootstrap & build phases, the state is stored and manipulated using the [Redux](https://redux.js.org/) library. The key purpose of using Redux in Gatsby internals is to centralize all of the state logic. Reviewing the Gatsby [reducers](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/reducers) and [actions](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/actions) folders gives a comprehensive picture of what state manipulations are possible.
```
Adding another nod to Gatsby Internals to make clear this is specific for Gatsby under the hood.

## Store

The namespaces in Gatsby's Redux store are a great overview of the Gatsby internals, here are a few:

- [Pages](https://www.gatsbyjs.org/docs/gatsby-internals-terminology/#redux-pages-namespace)
- [Components](https://www.gatsbyjs.org/docs/gatsby-internals-terminology/#redux-components-namespace)

> - Nodes

Reviewer: I wish there was more info for each of these store items, as they are really interesting parts of Gatsby's architecture. I feel like general descriptions of these would be a great outcome for this doc. If it's possible to describe them in original text here, that would be ideal; but we can also open issue(s) to add docs descriptions for them if necessary.

Author: Oof I enjoyed looking into these, really interesting. Hopefully this is roughly what you had in mind. Tried to keep them as short as possible, they could be more consistent in how I've structured each sentence maybe?

- Schema
- Jobs
- Webpack
- Metadata

> The Gatsby [/redux index file](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/index.ts) has two key exports, `store` and `emitter`. Throughout the bootstrap & build phases, `store` is used to get the current state & dispatch actions, while `emitter` is used to register listeners for particular actions. The store is also made available to Gatsby users through the [Node APIs](https://www.gatsbyjs.org/docs/node-apis/).

Reviewer 2: ```suggestion
The Gatsby [/redux index file](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/index.ts) has two key exports, `store` and `emitter`. Throughout the bootstrap and build phases, `store` is used to get the current state and dispatch actions, while `emitter` is used to register listeners for particular actions. The store is also made available to Gatsby users through the [Node APIs](https://www.gatsbyjs.org/docs/node-apis/).
```

> The Gatsby [/redux index file](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/index.ts) has two key exports, `store` and `emitter`. Throughout the bootstrap and build phases, `store` is used to get the current state and dispatch actions, while `emitter` is used to register listeners for particular actions. The store is also made available to Gatsby users through the [Node APIs](https://www.gatsbyjs.org/docs/node-apis/).

Reviewer: ```suggestion
The Gatsby [/redux index file](https://github.com/gatsbyjs/gatsby/tree/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/index.ts) has two key exports, `store` and `emitter`. Throughout the bootstrap and build phases, `store` is used to get the current state and dispatch actions, while `emitter` is used to register listeners for particular actions. The store is also made available to Gatsby users through the [Node APIs](/docs/node-apis/).
```

## Actions

Actions dispatched can have a series of effects as different reducers pick them up. Gatsby augments the typical Redux reducer with a pubsub [mitt](https://www.npmjs.com/package/mitt) `emitter` which subscribes to the Redux store to perform further side-effects.

> Actions dispatched in the store cause state changes through the reducers and also trigger listeners registered for that action on the `emitter`. The `subscribe` Redux store method is typically used to connect a web framework like React, Gatsby only uses the `subscribe` method to connect the `emitter`.

Reviewer: ```suggestion
Actions dispatched in the store cause state changes through the reducers and also trigger listeners registered for that action on the `emitter`. While the `subscribe` Redux store method is typically used to connect a web framework like React, Gatsby only uses the `subscribe` method to connect the `emitter`.
```

> The [Gatsby actions](/docs/actions/) are all either internal, public or restricted. The public actions, and a context relevant subset of the restricted actions, are available to users through the [Node APIs](https://www.gatsbyjs.org/docs/node-apis/). 

Reviewer: ```suggestion
The [Gatsby actions](/docs/actions/) are all either internal, public or restricted. The public actions, and a context relevant subset of the restricted actions, are available to users through the [Node APIs](/docs/node-apis/). 
```

> The Gatsby actions are all either internal, public or restricted. The public actions, and a context relevant subset of the restricted actions, are available to users through the [Node APIs](https://www.gatsbyjs.org/docs/node-apis/). There is a Gatsby docs page detailing all the public and restricted actions [here](/docs/actions/).

Reviewer 2: ```suggestion
The [Gatsby actions](/docs/actions/) are all either internal, public or restricted. The public actions, and a context relevant subset of the restricted actions, are available to users through the [Node APIs](https://www.gatsbyjs.org/docs/node-apis/). 
```

> ## Action Journey

Reviewer 2: Should this be a subheading of actions?

Author: Ah, yeh that would make sense.
```suggestion
### Action Journey
```

> At the time of writing there are 10 places in the code where `CREATE_NODE` type actions are dispatched, and 10 places where they are acted on through either reducers or `emitter` listeners.

Reviewer: I feel like this whole section is going to be outdated fast. Is there a way to write about it more generally that doesn't depend so closely on the source code as it's known now?

Author: Yeh, definitely. I've taken out the direct file/line references, made the opening sentence a little less specific and gone a little less deep in the binding step.

It's still dependent on the createRedirect action but I think any way it's written I'd be putting examples in. What do you think?

Reviewer: It's improved quite a bit from a maintainability standpoint as long as the createRedirect process doesn't change too much! I added some review suggestions to make it clear that it's an example.

> Here is the journey of a simpler public action, [createRedirect](/docs/actions/#createRedirect), through defining, exposing and dispatching:

Reviewer 2: ```suggestion
Here is the journey of the [createRedirect](/docs/actions/#createRedirect) public action through defining, exposing and dispatching:
```

> - **Reducer case** - [redirects.js](https://github.com/gatsbyjs/gatsby/blob/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/reducers/redirects.js#L28) - The redirects reducer will catch actions with a type `CREATE_REDIRECT` and makes the necessary state manipulation.

Reviewer 2: ```suggestion
- **Reducer case** - [redirects.js](https://github.com/gatsbyjs/gatsby/blob/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/reducers/redirects.js#L28) - The redirects reducer will catch actions with a type `CREATE_REDIRECT` and make the necessary state manipulation.
```

- **Action creator** - [public.js](https://github.com/gatsbyjs/gatsby/blob/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/redux/actions/public.js#L1358) - An action creator, `createRedirect`, is defined in the public actions file. The action has a payload, the information needed to complete the action, and a type, the string that identifies this particular action.

- **Expose bound action creator** - [api-runner-node.js](https://github.com/gatsbyjs/gatsby/blob/80acb8d5d67f7e277ce44158b36da84d262e5b23/packages/gatsby/src/utils/api-runner-node.js#L102) - `createRedirect` is one of the public actions made available to all of the [Node APIs](/docs/node-apis/). Node APIs are all run by the `apiRun` function found in `api-runner-node.js`. A collection of public actions and the restricted actions available to the called API are bound to the Redux store dispatch. The bound action collection is then passed when calling the user's API function.

> - **Dispatch** - Here is an example of the `createRedirect` call that a Gatsby user could make with the [createPages](https://www.gatsbyjs.org/docs/node-apis/#createPages) API in their project's [gatsby-node.js](https://www.gatsbyjs.org/docs/api-files-gatsby-node/) file:

Reviewer: ```suggestion
- **Dispatch** - Here is an example of the `createRedirect` call that a Gatsby user could make with the [createPages](/docs/node-apis/#createPages) API in their project's [gatsby-node.js](/docs/api-files-gatsby-node/) file:
```

> - **Dispatch** - Here an example `createRedirect` call that a Gatsby user could make with the [createPages](https://www.gatsbyjs.org/docs/node-apis/#createPages) API in their project's [gatsby-node.js](https://www.gatsbyjs.org/docs/api-files-gatsby-node/) file:

Reviewer 2: ```suggestion
- **Dispatch** - Here is an example of the `createRedirect` call that a Gatsby user could make with the [createPages](https://www.gatsbyjs.org/docs/node-apis/#createPages) API in their project's [gatsby-node.js](https://www.gatsbyjs.org/docs/api-files-gatsby-node/) file:
```
